### PR TITLE
http2: Proactively disconnect connections flooded when resetting stream

### PR DESCRIPTION
**Pull request by yanavlasov**
Commit Message:
Proactively disconnect connections flooded when resetting stream

Additional Description:
This PR adds error handling and associated tests for flood triggered in the ConnectionImpl::StreamImpl::resetStream() method. Previously connection stayed in the flooded state until downstream data was received.

Part of fixing #12280

Risk Level: Medium, H/2 codec
Testing: Unit and integration tests
Docs Changes: N/A
Release Notes: N/A

Signed-off-by: Yan Avlasov <yavlasov@google.com>

### CODE REVIEW

File: test/integration/autonomous_upstream.cc [hunk 1]
```diff
@@ -166,4 +169,16 @@ std::unique_ptr<Http::MetadataMapVector> AutonomousUpstream::preResponseHeadersM
   return std::move(pre_response_headers_metadata_);
 }
 
+AssertionResult AutonomousUpstream::closeConnection(uint32_t index,
+                                                    std::chrono::milliseconds timeout) {
+  if (!shared_connections_[index]->connected()) {
+    return AssertionSuccess();
+  }
```
Comment (mattklein123) [inline test/integration/autonomous_upstream.cc:176]: Technically can't this race? Should this part also be executed on the dispatcher and just ASSERT there?

**Review by mattklein123 — COMMENTED** [review]
Nice, one small question.

/wait-any

File: test/integration/autonomous_upstream.cc (hunk 1 above)
Comment (yanavlasov) [reply test/integration/autonomous_upstream.cc]: Sounds good.

**Comment by mattklein123** [comment]
OSX failure looks possibly legit.

/wait

**Comment by yanavlasov** [comment]
It looks like a flaky metadata test suite. There are other PRs that fail in the same ASSERT. Like [this PR](https://dev.azure.com/cncf/envoy/_build/results?buildId=53868&view=logs&j=a5e52b91-c83f-5429-4a68-c246fc63a4f7&t=5852bf5a-5a02-52f3-bee8-4fdc90cda9d0&l=5832)  I've filed issue #13527 to track this.

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**